moz/l10n/paths/discover.py

```diff
@@ -0,0 +1,235 @@
+# Copyright Mozilla Foundation
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from __future__ import annotations
+
+from os import sep, walk
+from os.path import commonpath, isdir, join, normpath, relpath, splitext
+from re import compile
+
+from moz.l10n.resource.format import l10n_extensions
+from moz.l10n.util import walk_files
+
+REF_DIR_SCORES = {
+    "templates": 3,
+    "en-US": 2,
+    "en-us": 2,
+    "en_US": 2,
+    "en_us": 2,
+    "en": 1,
+}
+
+locale_id = compile(
+    r"[a-z]{2}(?:[-_][A-Z][a-z]{3})?(?:[-_][A-Z]{2})?|ca-valencia|ja-JP-mac"
+)
+
+
+def dir_contains(dir: str, path: str) -> bool:
+    return commonpath((dir, path)) == dir
+
+
+def locale_dirname(base: str, locale: str) -> str:
+    if "-" in locale and not isdir(join(base, locale)):
+        alt_dir = locale.replace("-", "_")
+        if isdir(join(base, alt_dir)):
+            return alt_dir
+    return locale
+
+
+class MissingSourceDirectoryError(Exception):
+    """Raised when no source directory can be found."""
+
+
+class L10nDiscoverPaths:
+    """
+    Automagical localization resource discovery.
+
+    Given a root directory, finds the likeliest reference and target directories.
+
+    The reference directory has a name like `templates`, `en-US`, or `en`,
+    and contains files with extensions that appear localizable.
+
+    The localization target root is a directory with subdirectories named as
+    BCP 47 locale identifiers, i.e. like `aa`, `aa-AA`, `aa-Aaaa`, or `aa-Aaaa-AA`.
+
+    An underscore may also be used as a separator, as in `en_US`.
+    """
+
+    base: str | None
+    """The target base directory, with subdirectories for each locale."""
+    locales: list[str] | None
+    """Locales detected from subdirectory names under `base`."""
+    ref_paths: tuple[str, ...]
+    """Reference paths"""
+
+    def __init__(
+        self,
+        root: str,
+        ref_root: str | None = None,
+        ignorepath: str | None = ".l10n-ignore",
+    ) -> None:
+        """
+        To ignore files, include a `.l10n-ignore` file in the reference base directory
+        or some other location passed in as `ignorepath`.
+        This file uses git-ignore syntax,
+        and is always based in the reference base directory.
```

Comment:
What does _based in_ mean? Located in?

Reply:
Means that it's the root directory for the expressions, even if the `.l10n-ignore` file is itself at a different path.